File: src/main/java/pl/patrykdolata/lifemanager/controller/web/WebExceptionHandler.java
package pl.patrykdolata.lifemanager.controller.web;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.AuthenticationException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
import pl.patrykdolata.lifemanager.exceptions.WebException;
import pl.patrykdolata.lifemanager.model.ErrorResponse;
import pl.patrykdolata.lifemanager.util.ResponseUtils;

import javax.servlet.http.HttpServletRequest;
import java.util.Date;

@ControllerAdvice
public class WebExceptionHandler extends ResponseEntityExceptionHandler {

    @ExceptionHandler(AuthenticationException.class)
    public final ResponseEntity<ErrorResponse> handleAuthenticationExceptions(Exception ex, HttpServletRequest request) {
        return ResponseUtils.response(createErrorResponse(ex, HttpStatus.UNAUTHORIZED.value(), request),
                HttpStatus.UNAUTHORIZED);
    }

    @ExceptionHandler(WebException.class)
    public final ResponseEntity<ErrorResponse> handleWebExceptions(WebException ex, HttpServletRequest request) {
        return ResponseUtils.response(createErrorResponse(ex, ex.getStatusCode().value(), request),
                ex.getStatusCode());
    }

    private ErrorResponse createErrorResponse(Exception ex, int status, HttpServletRequest request) {
        return new ErrorResponse(ex.getMessage(), status, new Date().getTime(), request.getRequestURI());
    }
}

File: src/main/kotlin/pl/patrykdolata/lifemanager/controller/web/WebExceptionHandler.kt
package pl.patrykdolata.lifemanager.controller.web

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.AuthenticationException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
import pl.patrykdolata.lifemanager.exceptions.WebException
import pl.patrykdolata.lifemanager.model.ErrorResponse
import pl.patrykdolata.lifemanager.util.ResponseUtils.response
import java.util.*
import javax.persistence.EntityNotFoundException
import javax.servlet.http.HttpServletRequest

@ControllerAdvice
class WebExceptionHandler : ResponseEntityExceptionHandler() {

    @ExceptionHandler(AuthenticationException::class)
    fun handleAuthenticationExceptions(ex: Exception, request: HttpServletRequest): ResponseEntity<ErrorResponse> =
            response(createErrorResponse(ex.message!!, HttpStatus.UNAUTHORIZED.value(), request), HttpStatus.UNAUTHORIZED)

    @ExceptionHandler(WebException::class)
    fun handleWebExceptions(ex: WebException, request: HttpServletRequest): ResponseEntity<ErrorResponse> =
            response(createErrorResponse(ex.message!!, ex.statusCode.value(), request), ex.statusCode)

    @ExceptionHandler(EntityNotFoundException::class)
    fun handleEntityNotFoundException(ex: EntityNotFoundException, request: HttpServletRequest)
            : ResponseEntity<ErrorResponse> {
        val message = ex.message ?: "Entity not found"
        return response(createErrorResponse(message, HttpStatus.NOT_FOUND.value(), request), HttpStatus.NOT_FOUND)
    }

    private fun createErrorResponse(message: String, status: Int, request: HttpServletRequest): ErrorResponse =
            ErrorResponse(message, status, Date().time, request.requestURI)
}
